### ISSUE TITLE
Repository reported as containing HTML but the majority of files are Scheme

### ISSUE THREAD

**Issue by NewForester** [github/linguist#3554]
Hi,

Repository:  https://github.com/NewForester/fpl-scheme.

Summary reports that is contains HTML.  In IMHO this is a misrepresentation.

The Language Bar reports 67.0% of files are HTML, 32.8% are Scheme and 2% are CSS.  This, IMHO, is plain wrong.

The breakdown, according to the UI, is:
  * 46 Scheme files
  * 20 HTML files
  * 20 Markdown files
  * 1 CSS file
  * 1 txt file

These tallies, IMHO, are correct.

According to my pre-Brexit but post-Imperial arithmetic, that means 52.27% of files are Scheme and even if Markdown files count as HTML, Scheme is still the democratic majority.  IMHO the repository should report itself as containing Scheme.

What have I done wrong ?

**Comment by larsbrinkhoff** [comment]
The percentages are calculated according to file sizes, not the number of files.

**Comment by Alhadis** [comment]
Linguist calculates usage stats based on filesize, not their head count.

If you click the language bar, you'll see see the [language breakdown](https://github.com/NewForester/fpl-scheme/search?l=html) for your codebase: each of those HTML files is skewing the repository's classification.

Thankfully, there's a [documented workaround](https://github.com/github/linguist/#overrides). Simply add the following to a `.gitattributes` file in the root of your repository with the following lines:

~~~
/html linguist-documentation
/LICENCE.html linguist-documentation
~~~

This instructs Linguist to exclude them from language stats... there's more info on the link I gave you.

**Comment by Alhadis** [comment]
... ffs Lars, JUST as I was finishing... :(

**Comment by larsbrinkhoff** [comment]
Sorry, nice guys finish last.

**Comment by Alhadis** [comment]
Well I'm the angry, tactless prick of the group, so I should've finished first. 🥇

**Comment by pchaigno** [comment]
@Alhadis If you want to race him, you could use saved replies :zap:

**Comment by Alhadis** [comment]
I prefer to compete with a 150+ wpm typing speed.

Then again, it probably didn't help that my posture was so bad, my chin has handling more keystrokes than my hands were.

**Comment by NewForester** [comment]
Thankyou @Alhadis for the .gitattributes suggestion.  Most grateful.

I tried it on the repository in question.   Nothing happen so I left it 24 hours to see if the files concerned would be re-scanned or whatever. Seems not.

Perhaps, I thought, the scanning only occurs when changes are pushed to GitHub and only files that have been changed are scanned.

I pushed the repository into one I keep for messing around with github (https://github.com/NewForester/quick-trial).  The stats showed this clone contained mostly HTML.

I then removed the HTML files and pushed a new commit.  The effect was immediate:  the repository was reported as containing Scheme.  Overwhelmingly so.  It seems Markdown does not count for anything.

I then reverted the commit and pushed again.  Sadly the effect was to revert the language statistics too.  The work around does not seem to do what I inferred you implied it would.

What else do I not yet know ?

**Comment by Alhadis** [comment]
Ah damn, I always forget to mention the caching issue.

GitHub caches usage statistics until the files are updated again: it doesn't update them whenever a `.gitattributes` rule is updated with a `linguist-*` property (although it probably should...).

**Easiest fix:** just push a meaningless change to an existing HTML file, and that'll force a reindexing of your repository. =)

**Comment by NewForester** [comment]
Thanks.  I guessed there was a caching issue but it seems that it not the whole story.

I tried as you suggested by altering one of the HTML files in the subdirectory the `.gitattributes` file says linguist is to ignore (that is linguist with a g not a q ?).  No joy when I pushed it.

I tried altering the HTML file that is not in a subdirectory and is not to be ignored.  No joy either.

I had earlier changed `.gitattributes` to request the directory containing the only CSS file be ignored.  I tried changing the CSS file and pushing the commit.  Still no joy.

Here is a copy paste of:

```bash
work:quick-trial$ cat .gitattributes 
/html   linguist-documentation
/styles linguist-documentation
```

I think everything (even the file name) is spelt the way it should be.

GitHub and Linguist are not my code so it is mine to wonder and, perhaps, yours to know. If it were my code I would be begining to wonder if it had not regressed.

**Comment by Alhadis** [comment]
Ugh, sorry, it's my fault. It should've been this:

~~~
html/*   linguist-documentation
styles/* linguist-documentation
~~~

Not sure why `/html` wasn't working. From the **gitattributes**(8) manpage:

> The rules how the pattern matches paths are the same as in **.gitignore** files; see
> **gitignore**(5). Unlike **.gitignore**, negative patterns are forbidden.

Guess GitHub (or Linguist) isn't using Git to parse these patterns...

**Comment by NewForester** [comment]
Ah, that did the trick and without any shenanigans to flush the GitHub cache.

This is not your fault but a documentation error.  One down from user error in my experience.  At least my problem is resolved and my tail is not between my legs so I call that a win.  This is not the first time someone has claimed 'just like in git' and it isn't.  I wonder if they know what a git is.

The documentation in this area is poor.  That is ... so scant that anyone reading it will find it difficult not to read between the commits what they want to.

The GitHub help has only the brief page: [About repository languages](https://help.github.com/articles/about-repository-languages/).  It makes no mention of how the statistics are calculated, merely implying that the only reason you are reading the page is because they are wrong and it passes the buck to the [linguist readme](https://github.com/github/linguist#troubleshooting) file.

GitHub help does not even tell you how to find the languages statistics.  I guess it is intuitive that a systematic search of the UI will yield nothing but that a pause of a second after moving the mouse at random will eventually produce the tool-tip that reads "Click for language details".

The linguist read me is not incorrect but it does not refer to lines of code but "bytes of code"  which, I kid you not, puzzled me because I thought git and GitHub were source code repositories and not about the compiled stuff.

I took it from the context that this analysis was about deciding not to count binary blobs and stuff.  That is about deciding which files not to count so, by Occams razor, I took it that the statistics were counts of files that it decided to count.  Now that I know it isn't see that the documentation does not imply this.

This proves the documentation error point.  Documentation is not useless because no one reads it.  It is useless because no one who knows will read it (including the git who wrote it) and anyone who does not will not be able to make sense of it.

Anyway, enjoy your evening or whatever it is in your timezone.

P.S.  I'm a 140 words an hour man.

P.P.S  What is the etiquette now ?  Should I close this issue before somegit marks it as invalid ?
